src/hotspot/share/opto/regmask.hpp

```diff
@@ -260,115 +548,299 @@ class RegMask {
   static int num_registers(uint ireg);
   static int num_registers(uint ireg, LRG &lrg);
 
-  // Fast overlap test.  Non-zero if any registers in common.
+  // Overlap test. Non-zero if any registers in common, including all-stack.
   bool overlap(const RegMask &rm) const {
+    assert(_offset == rm._offset, "offset mismatch");
     assert(valid_watermarks() && rm.valid_watermarks(), "sanity");
+
+    // Very common overlap case: _rm_up overlap. Check first to reduce
+    // execution time.
     unsigned hwm = MIN2(_hwm, rm._hwm);
     unsigned lwm = MAX2(_lwm, rm._lwm);
-    uintptr_t result = 0;
     for (unsigned i = lwm; i <= hwm; i++) {
-      result |= _RM_UP[i] & rm._RM_UP[i];
+      if (_rm_up(i) & rm._rm_up(i)) {
+        return true;
+      }
+    }
+
+    // Very rare overlap cases below.
+
+    // We are both all-stack
+    if (is_AllStack() && rm.is_AllStack()) {
+      return true;
     }
-    return result;
+
+    // We are all-stack and rm _hwm is bigger than us
+    if (is_AllStack() && rm._hwm >= _rm_size) {
+      for (unsigned i = MAX2(rm._lwm, _rm_size); i <= rm._hwm; i++) {
+        if (rm._rm_up(i)) {
+          return true;
+        }
+      }
+    }
+
+    // rm is all-stack and our _hwm is bigger than rm
+    if (rm.is_AllStack() && _hwm >= rm._rm_size) {
+      for (unsigned i = MAX2(_lwm, rm._rm_size); i <= _hwm; i++) {
+        if (_rm_up(i)) {
+          return true;
+        }
+      }
+    }
```

Comment:
There is a bit of code duplication here. A helper method could help, no pun intended.

Reply:
Boah, but it is also not horrible to leave as is. Maybe a helper method does not make more readable. Not sure.

Reply:
I see your point, but I'm not sure how much it helps readability. I suggest leaving it as is.